### DISCUSSION TITLE
Interaction with Prometheus Server

### DISCUSSION THREAD

**Discussion by sharmila359** [open-telemetry/opentelemetry-cpp#559] [General]
What kind of change we have to do to hit Prometheus server from Exporter ???
Like code change , config change , address & port setting etc
Any sample code or GitHub reference will be very helpful.

Awaiting for quick response........

Thanks in advanced.

**Comment by lalitb** [comment]
Current Prometheus exporter implementation collects the metric data in format expected by Prometheus server, but doesn't really do any export. Export functionality is not present for now. Also, as the current metrics spec is experimental, these changes are not planned as of now.

**Comment by sba429** [comment]
Thanks for you input. Now let me share the output..we have 
Prometheus Exporter instance running and exposes an HTTP endpoint at localhost:8080 and prometheus server running at localhost:9090
now we are using config as follows,
scrape_configs:
  - job_name: 'prometheus'
    scrape_interval: 5s
    static_configs:
      - targets: ['localhost:9090']

  - job_name: 'mpctest'
    scrape_interval: 5s
    static_configs:
      - targets: ['localhost:8080']
Now when we are running this we can see some failures as follows and nothing is getting displayed in graph. Please have a look and provide your input please.
# HELP net_conntrack_dialer_conn_established_total Total number of connections successfully established by the given dialer a given name.
# TYPE net_conntrack_dialer_conn_established_total counter
net_conntrack_dialer_conn_established_total{dialer_name="default"} 0
net_conntrack_dialer_conn_established_total{dialer_name="mpctest"} 0
net_conntrack_dialer_conn_established_total{dialer_name="prometheus"} 1
# HELP net_conntrack_dialer_conn_failed_total Total number of connections failed to dial by the dialer a given name.
# TYPE net_conntrack_dialer_conn_failed_total counter
net_conntrack_dialer_conn_failed_total{dialer_name="default",reason="refused"} 0
net_conntrack_dialer_conn_failed_total{dialer_name="default",reason="resolution"} 0
net_conntrack_dialer_conn_failed_total{dialer_name="default",reason="timeout"} 0
net_conntrack_dialer_conn_failed_total{dialer_name="default",reason="unknown"} 0
net_conntrack_dialer_conn_failed_total{dialer_name="mpctest",reason="refused"} 139
net_conntrack_dialer_conn_failed_total{dialer_name="mpctest",reason="resolution"} 0
net_conntrack_dialer_conn_failed_total{dialer_name="mpctest",reason="timeout"} 0
net_conntrack_dialer_conn_failed_total{dialer_name="mpctest",reason="unknown"} 139
net_conntrack_dialer_conn_failed_total{dialer_name="prometheus",reason="refused"} 0
net_conntrack_dialer_conn_failed_total{dialer_name="prometheus",reason="resolution"} 0
net_conntrack_dialer_conn_failed_total{dialer_name="prometheus",reason="timeout"} 0
net_conntrack_dialer_conn_failed_total{dialer_name="prometheus",reason="unknown"} 0

**Comment by Krishna-Koley17** [comment]
We have added a new PrometheusExporter class like below :-



class PrometheusExporter //: public metric_sdk::MetricsExporter
{
public:
       PrometheusExporter(const std::string &);
       void Export(std::vector<metric_sdk::Record> &);
       //metric_sdk::ExportResult Export(const std::vector<metric_sdk::Record> & obj);
       void shutdown();
       void getCollector();
       bool isShutdown();

private:
      std::shared_ptr<PrometheusCollector> collector;
      std::unique_ptr<Exposer> exposer;
      bool is_shutdown;
};

And Export() method is added like below :-
void PrometheusExporter::Export(std::vector<metric_sdk::Record> & obj)
{
   std::cout<<" \n PrometheusExporter::Export method is called ";
   std::cout<<" \n Record details :";
   for(std::vector<metric_sdk::Record>::iterator iter = obj.begin();iter != obj.end();++iter)
   {
       std::cout<<"\n Name : "<<iter->GetName()<<" Description :"<<iter->GetDescription()<<" Labels:"<<iter->GetLabels();
   }
   collector->AddMetricData(obj);
 
}

And from our main() method we called Export() method below 
metric_sdk::Record get_record(const std::string &type,
                              int version,
                              const std::string &label,
                              std::shared_ptr<metric_sdk::Aggregator<T>> aggregator)
{
  std::string name = "test_" + type + "_metric_record_v" ;//+ std::to_string(version) + ".0";
  std::string desc = "this is a test " + type + " metric record";
  Record record(name, desc, label, aggregator);
  return record;
}

int main()
{
    // declare HTTP endpoint to expose, and start a Prometheus exporter
    std::string address = "10.115.49.155:5055";
    PrometheusExporter exporter{address};

std::vector<Record>collection;
    std::vector<double> boundaries{10, 20, 30, 40, 50};
    auto aggregator = std::shared_ptr<Aggregator<int>>(new HistogramAggregator<int>(opentelemetry::metrics::InstrumentKind::Counter, boundaries));

 Record record = get_record(arr[0],1,arr[2], aggregator);
               aggregator->checkpoint();
               collection.emplace_back(record);
               exporter.Export(collection);

               std::this_thread::sleep_for(std::chrono::milliseconds(2500));



But we are not able to see those metric details in the metric page of Prometheus server. Our Prometheus server is running in http://localhost:9090/ and Prometheus Exporter is running in http://localhost:5050

Our metric data should start with "test_" as in get_record(..) method it is concatenated with "test_".

Please suggest why we are not able to see our metric data in Prometheus/metric page.
